File: proga/1_semestr/labs/lab3/src/main/java/dev/lysmux/lab3/story/Story.java
package dev.lysmux.lab3.story;

import dev.lysmux.lab3.story.exception.EmptyStoryException;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class Story {
    protected List<Sentence> sentences;

    public Story() {
        sentences = new ArrayList<Sentence>();
    }

    public void addSentences(Sentence... sentences) {
        this.sentences.addAll(List.of(sentences));
    }

    public List<Sentence> getSentences() {
        return sentences;
    }

    public String makeStory() {
        if (sentences.isEmpty()) throw new EmptyStoryException();

        return sentences.stream().map(Sentence::content)
                .collect(Collectors.joining(". "));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Story story)) return false;
        return Objects.equals(sentences, story.sentences);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(sentences);
    }

    @Override
    public String toString() {
        return "Story{" +
                "sentences=" + sentences +
                '}';
    }
}

File: proga/1_semestr/labs/lab3/src/main/kotlin/dev/lysmux/lab3/story/Story.kt
package dev.lysmux.lab3.story

import dev.lysmux.lab3.story.exception.EmptyStoryException


class Story {
    private val sentences = ArrayList<Sentence>()

    companion object {
        fun build(init: Story.() -> Unit) = Story().apply(init)
    }

    fun sentence(init: SentenceBuilder.() -> Unit) {
        sentences.add(SentenceBuilder().apply(init).build())
    }

    fun asText(): String {
        if (sentences.isEmpty()) throw EmptyStoryException()
        return sentences.joinToString(". ") { it.content }
    }
}
